File: M1/azor-chatdog-py/src/files/wal.py
import os
import json
from datetime import datetime
from files.config import WAL_FILE


def append_to_wal(session_id: str, prompt: str, response_text: str, total_tokens: int, model_name: str) -> tuple[bool, str | None]:
    """
    Appends a transaction to the WAL (Write-Ahead Log) file.
    
    Args:
        session_id: Unique session identifier
        prompt: User's input prompt
        response_text: Model's response text
        total_tokens: Total tokens used in the conversation
        model_name: Name of the LLM model used
    
    Returns:
        tuple[bool, str | None]: (success, error_message)
    """
    
    wal_entry = {
        'timestamp': datetime.now().isoformat(),
        'session_id': session_id,
        'model': model_name,
        'prompt': prompt,
        'response': response_text,
        'tokens_used': total_tokens
    }

    try:
        if os.path.exists(WAL_FILE) and os.path.getsize(WAL_FILE) > 0:
            with open(WAL_FILE, 'r', encoding='utf-8') as f:
                try:
                    data = json.load(f)
                except json.JSONDecodeError:
                    data = []
                    return False, f"WAL file corrupted, resetting: {WAL_FILE}"
        else:
            data = []

        data.append(wal_entry)

        with open(WAL_FILE, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=4, ensure_ascii=False)
            return True, None
            
    except Exception as e:
        return False, f"Error writing to WAL file ({WAL_FILE}): {e}"

File: M1/azor-chatdog-js/src/files/wal.ts
/**
 * Write-Ahead Log (WAL) for transaction audit trail
 */

import fs from 'fs';
import { getWALFile } from './config.js';
import type { WALEntry, Result } from '../types/index.js';

/**
 * Append an entry to the Write-Ahead Log
 */
export function appendToWAL(
  sessionId: string,
  prompt: string,
  responseText: string,
  totalTokens: number,
  modelName: string
): Result<boolean, string> {
  const walFile = getWALFile();

  const entry: WALEntry = {
    timestamp: new Date().toISOString(),
    session_id: sessionId,
    model: modelName,
    prompt,
    response: responseText,
    tokens_used: totalTokens,
  };

  try {
    // Read existing WAL if it exists
    let walData: WALEntry[] = [];
    if (fs.existsSync(walFile)) {
      try {
        const content = fs.readFileSync(walFile, 'utf-8');
        walData = JSON.parse(content);
      } catch {
        // If WAL is corrupted, start fresh
        walData = [];
      }
    }

    // Append new entry
    walData.push(entry);

    // Write back to file
    fs.writeFileSync(walFile, JSON.stringify(walData, null, 2), 'utf-8');

    return { success: true, value: true };
  } catch (error) {
    return {
      success: false,
      error: `Error writing to WAL: ${(error as Error).message}`,
    };
  }
}

/**
 * Read all WAL entries
 */
export function readWAL(): Result<WALEntry[], string> {
  const walFile = getWALFile();

  if (!fs.existsSync(walFile)) {
    return { success: true, value: [] };
  }

  try {
    const content = fs.readFileSync(walFile, 'utf-8');
    const data = JSON.parse(content);
    return { success: true, value: data };
  } catch (error) {
    return {
      success: false,
      error: `Error reading WAL: ${(error as Error).message}`,
    };
  }
}
